src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/PropertySymbolReferenceFinder.cs

```diff
@@ -147,79 +142,83 @@ private static bool IsForEachProperty(IPropertySymbol symbol)
-            var indexerMemberCrefs = syntaxRoot.DescendantNodes(descendIntoTrivia: true)
-                                               .Where(syntaxFacts.IsIndexerMemberCRef);
+            if (syntaxFacts.IsElementAccessExpression(node))
+            {
+                // For an ElementAccessExpression the indexer we are looking for is the argumentList component.
+                syntaxFacts.GetPartsOfElementAccessExpression(node, out var expression, out indexerReference);
 
-            var locations = ArrayBuilder<FinderLocation>.GetInstance();
+                if (expression != null && symbolsMatch(expression, semanticModel).matched)
+                {
+                    // Element access with explicit member name (allowed in VB).
+                    // We have already added a reference location for the member name identifier, so skip this one.
+                    return (indexerReference, matched, reason);
+                }
 
-            foreach (var node in indexerMemberCrefs)
+                (matched, reason) = symbolsMatch(node, semanticModel);
+            }
+            else if (syntaxFacts.IsConditionalAccessExpression(node))
             {
-                cancellationToken.ThrowIfCancellationRequested();
+                // For a ConditionalAccessExpression the whenNotNull component is the indexer reference we are looking for
+                syntaxFacts.GetPartsOfConditionalAccessExpression(node, out var expression, out indexerReference);
 
-                var (matched, reason) = symbolsMatch(node, semanticModel);
-                if (matched)
+                if (syntaxFacts.IsInvocationExpression(indexerReference) && symbolsMatch(indexerReference, semanticModel).matched)
                 {
-                    var location = node.SyntaxTree.GetLocation(new TextSpan(node.SpanStart, 0));
-                    var symbolUsageInfo = GetSymbolUsageInfo(node, semanticModel, syntaxFacts, semanticFacts, cancellationToken);
-                    locations.Add(new FinderLocation(
-                        node, new ReferenceLocation(document, null, location, isImplicit: false, symbolUsageInfo, GetAdditionalFindUsagesProperties(node, semanticModel, syntaxFacts), candidateReason: reason)));
+                    // Element access with explicit member name (allowed in VB).
+                    // We have already added a reference location for the member name identifier, so skip this one.
+                    return (indexerReference, matched, reason);
                 }
+
+                (matched, reason) = symbolsMatch(indexerReference, semanticModel);
+            }
+            else
+            {
+                Debug.Assert(syntaxFacts.IsIndexerMemberCRef(node));
+
+                // For an IndexerMemberCRef the node itself is the indexer we are looking for.
+                indexerReference = node;
+
+                (matched, reason) = symbolsMatch(node, semanticModel);
```

Comment:
can you extract each case of this if/elseif/else block into its own method?  It will def make this easier to read and understand.